src/Processors/QueryPlan/SortingStep.cpp

```diff
@@ -239,11 +242,70 @@ void SortingStep::finishSorting(
         });
 }
 
+void SortingStep::enableVirtualRow(const QueryPipelineBuilder & pipeline) const
```

Comment:
This is a wrong place to check whether we can enable virtual row or not.
Check should happen on the call side, every time we call `requestReadingInOrder`.
We can enable a virtual row if all the underlying plan steps can support it. Most likely it is true, we need to support FilterStep and ExpressionStep only. Other cases (DistinctStep, ArrayJoinStep, expressions with arrayJoin function) should disable this flag.

ExpressionTransfrom should only check an exception for a virtual row chunk, and ignore the chunk if the exception was thrown. FilterTransform additionally should **not** filter out the virtual row.

Right now filter transform can filter a virtual row; it is not a bug, but I think I can reproduce it.